File: frontend/src/api/types/tos.ts
export interface ToSRequest {
	is_html: boolean;
	opt_until?: number;
	content: string;
}

export interface ToSUserAcceptRequest {
	tos_ts: number;
	// 65 char code
	accept_code: string;
}

export interface ToSResponse {
	ts: number;
	author: string;
	is_html: boolean;
	opt_until?: number;
	content: string;
}

export interface ToSAwaitLoginResponse {
	tos_await_code: string;
	user_id: string;
}

export interface ToSLatestResponse {
	ts: number;
	is_html: boolean;
	opt_until?: number;
	content: string;
}

export interface ToSUserAcceptResponse {
	user_id: string;
	tos_ts: number;
	accept_ts: number;
	location: string;
}

File: src/api_types/src/tos.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use validator::Validate;

#[derive(Deserialize, ToSchema, Validate)]
pub struct ToSRequest {
    pub is_html: bool,
    #[validate(length(min = 1, max = 8192))]
    pub content: String,
}

#[derive(Deserialize, ToSchema, Validate)]
pub struct ToSAcceptRequest {
    pub tos_ts: i64,
    #[validate(length(min = 64, max = 64))]
    pub accept_code: String,
}

#[derive(Serialize, ToSchema)]
pub struct ToSResponse {
    pub ts: i64,
    pub author: String,
    pub is_html: bool,
    pub content: String,
}

#[derive(Serialize, ToSchema)]
pub struct ToSLatestResponse {
    pub ts: i64,
    pub is_html: bool,
    pub content: String,
}

#[derive(Serialize, ToSchema)]
pub struct ToSUserAcceptResponse {
    pub user_id: String,
    pub tos_ts: i64,
    pub accept_ts: i64,
    pub location: String,
}
